### PR TITLE
feat: add add-sequence route

### PR DESCRIPTION
**Pull request by ismay**
Note: there is no link to the added page (http://localhost:3000/#/add-sequence) in the ui yet because the list and edit routes haven't been added yet. We can expose the functionality once we've added the edit and list routes.

- https://dhis2.atlassian.net/browse/DHIS2-14077
- https://dhis2.atlassian.net/browse/DHIS2-14748
- https://dhis2.atlassian.net/browse/DHIS2-15556
- https://dhis2.atlassian.net/browse/LIBS-515
- https://dhis2.stoplight.io/docs/dhis2/ehaxt2lzojavb-dhis-2-api-scheduler-only
- https://www.figma.com/file/YVXIsd80IgZnTfMDE0qOGf/Scheduler?type=design&node-id=215-2142
- https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-240/scheduling.html?h=2.40#queues-add

First

- [x] Handle loading or error in sequence order field
- [x] Handle selection in sequence order field
- [x] Set filter field placeholder text
- [x] Transform sequence queue field values to format expected by backend
- [x] Choose route name (add-sequence)

Validation

- [x] Set correct validators for sequence queue field
- [x] Show title
- [x] Show required status
- [x] Show states and helptext

Submitting

- [x] Create hook to submit sequence (and add to form)

Other

- [x] Create generic name form field, if BE requirements are the same
- [x] Add missing tests
  - [x] use-queueables
  - [x] sequence add form
  - [x] sequence add form container
  - [x] sequence order field
- [x] Use custom options for sequence order field

Follow-up

- No easy way to handle focus and blur with the transfer, so validation only happens on submission and not also on blur like the other form elements (this is similar to the other form elements in this app that use the transfer) (https://dhis2.atlassian.net/browse/LIBS-97, https://dhis2.atlassian.net/browse/LIBS-123, https://dhis2.atlassian.net/browse/LIBS-160, https://dhis2.atlassian.net/browse/LIBS-163)
- Custom transfer option is not accessible at all (maybe standard transfer option isn't either), we should fix that (see issues in point above this one, this also falls under those issues).
- Move history.push out of hooks and to form submit handlers (https://dhis2.atlassian.net/browse/DHIS2-15476)
- Split up the other complex form fields that are similar to the sequence order field in complexity (https://dhis2.atlassian.net/browse/DHIS2-15477)

### CODE REVIEW

File: src/components/FormFields/SequenceOrderField/SequenceTransfer.js
Comment (ismay) [reply src/components/FormFields/SequenceOrderField/SequenceTransfer.js:22]: Hmm maybe that's because I'm using a custom option. Could be that they require a little extra work to support that. I'll look into it.

File: src/components/Forms/SequenceAddForm.test.js
Comment (ismay) [reply src/components/Forms/SequenceAddForm.test.js:1]: > We talked before about the approach of ["write tests. Mostly integration"](https://kentcdodds.com/blog/write-tests)

The way I read this article the core message is: testing is valuable, and devs should keep cost/benefit in mind when writing tests. I agree with that. Though I personally do not see a lot of value in the article and the tweet it's based beyond that. I definitely agree that the tests for this app could be improved. The tests I wrote in this PR are in keeping with the existing test style in this repo, but I think there are several areas that could use some attention if we're going to revamp the approach:

- The mocking is decreasing test realism. I would have to look through the app to see why exactly we're mocking, but I think it's probably often a composition/architecture issue. i.e., the test just wants to assert something about a component, but the component includes a lot of other components/logic that requires too much setup, so it's mocked to contain the scope. Using composition instead of hardcoding the structure would probably resolve the need for mocking without a lot of added complexity.
- Accessibility. In terms of realism simple jest unit/integration tests are often hindered by our component lib not being very accessible. So instead of being able to use react testing lib's recommended selectors we're often using data-test attr's and the like. Brittle, complex and not very realistic.
- App architecture. The dynamic nature of the app also introduces a lot more logic. I think moving to a more declarative architecture will also allow us to simplify our test suite a lot. The parameter options component wouldn't be necessary for example, which is now one of our more complex components.
- Visual regression tests. We're now asserting things that are more efficiently asserted visually. Moving to visual regression tests would save devs a fair bit of time I think.

In terms of the architecture I think this app also still contains a couple of redux conventions that could probably be refactored. That's where the containers come from. In terms of integration, unit and e2e, I don't have an ideological preference or aversion. Whatever keeps the codebase cost-effective, reliable and easy to change and maintain.

File: cypress/integration/add-sequence/back-to-all-jobs.feature
Comment (ismay) [reply cypress/integration/add-sequence/back-to-all-jobs.feature:1]: It's not a _super_ expensive test in my opinion. It's easy to maintain and runs in about two seconds. But we could remove it. In general we could look at which paths we want to test as e2e tests and which we don't. Thinking a bit about that would probably be good, because this test suite has just been growing organically for a while.

File: cypress/fixtures/add-sequence/two-unqueued-jobs.json
Comment (ismay) [reply cypress/fixtures/add-sequence/two-unqueued-jobs.json:1]: I would say so yes. I think the fixtures have the benefit of being stable, but that can also be a downside in that it decreases the realism. The running cost of a dockerized backend is not that much higher, and it's less complex in my opinion. So yeah, I would be in favor.

File: src/components/FormFields/SequenceOrderField/SequenceTransfer.js
Comment (ismay) [reply src/components/FormFields/SequenceOrderField/SequenceTransfer.js:22]: I've added the doubleclick handler. I noticed that in our native transfer options we don't use the onDoubleClick prop, but instead a custom solution: https://github.com/dhis2/ui/blob/master/components/transfer/src/transfer-option.js. We should probably swap that out. Also, I think it's a slightly confusing pattern personally, but since we already employ the pattern elsewhere it makes sense to at least have it here as well.

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js:19]: I was thinking that it might be good to first finish the edit form and the list and then take a look at cleaning up these kinds of patterns. There's a fair bit of cleanup to do I think, maybe getting the features out of the door first would be a good initial goal? Let me know what you think.

File: src/components/FormFields/SequenceOrderField/SequenceTransfer.js
Comment (ismay) [reply src/components/FormFields/SequenceOrderField/SequenceTransfer.js:22]: Yeah as an established pattern for our apps it would be weird to not have it here.

File: src/components/Forms/SequenceAddForm.test.js
Comment (ismay) [reply src/components/Forms/SequenceAddForm.test.js:1]: Yeah I agree with not mocking where possible. About the visual regression tests, I've used percy quite a bit, and for me it's been a good experience: https://percy.io. I've used it in combination with cypress and overall it's been a timesaver for me.

> Anyhow in the context of this PR, I think it's fair to just follow the existing patterns. We can tackle improving the tests separately as part of [the ticket](https://dhis2.atlassian.net/browse/DHIS2-15334) you added (and before the refactor imo)

Yeah that sounds good to me 👍

File: cypress/integration/add-sequence/back-to-all-jobs.feature
Comment (ismay) [reply cypress/integration/add-sequence/back-to-all-jobs.feature:1]: All right, sounds good to me 👍

File: src/components/FormFields/SequenceOrderField/SequenceTransfer.js
Comment (ismay) [reply src/components/FormFields/SequenceOrderField/SequenceTransfer.js:22]: Created an issue for fixing this in the transfer: https://dhis2.atlassian.net/browse/LIBS-515

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js:19]: For future reference, it's this issue: https://dhis2.atlassian.net/browse/DHIS2-15476. I'll work on that next.

File: src/components/Forms/SequenceAddForm.test.js
Comment (ismay) [reply src/components/Forms/SequenceAddForm.test.js:1]: Since we have the followup ticket (https://dhis2.atlassian.net/browse/DHIS2-15334) I'll close this as resolved.

File: cypress/integration/add-sequence/back-to-all-jobs.feature
Comment (ismay) [reply cypress/integration/add-sequence/back-to-all-jobs.feature:1]: I'll resolve this one as well, since we have a follow up ticket here as well.

File: cypress/fixtures/add-sequence/two-unqueued-jobs.json
Comment (ismay) [reply cypress/fixtures/add-sequence/two-unqueued-jobs.json:1]: I assume we'll get to this when the scheduler tickets are resolved and we can resume our work with the docker/e2e testing group.

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (tomzemp) [inline src/hooks/job-queues/use-submit-job-queue.js:29]: do you want to throw this error? It seems the error needs need to be returned to be be picked up by the react final form?

File: src/components/FormFields/SequenceOrderField/SequenceTransferTitle.js
Comment (tomzemp) [inline src/components/FormFields/SequenceOrderField/SequenceTransferTitle.js:4]: is there a reason to put the style here in the component rather than in a css module?

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (tomzemp) [inline src/hooks/job-queues/use-submit-job-queue.js:25]: FYI: when I got a 409 back from the api (due to entering the same name for a job sequence), the formatError function did not work because there was no property `response` on details

**Review by tomzemp — APPROVED** [review]
Looks good to me 🎉!

I didn't have any very specific issues with the code, but I did notice:

- I don't see errors coming through in the UI. Do you need to subscribe to `hasSubmitErrors` `submitErrors` in the Form component in SequenceAddFormContainer? I'm a bit confused by the react-final-form [docs](https://final-form.org/docs/react-final-form/types/FormRenderProps) because they seem to suggest you need to subscribe, but the `submitting` prop seems to be working without subscription 🤔
- I'm not sure if there is a plan to deal with the overall list page separately, but currently after the sequence gets added, when you click on edit, it will open the edit menu up for the first job in the sequence. Also if you click delete, it seems to just delete the first job in the sequence. 

_UI questions (so they might not be relevant if this has been decided elsewhere):_ I was a little confused by the fact that I had to add jobs, then I had to add a new sequence and that adding to the sequence would remove the job from the list. I don't know if this is something that would be addressed in the transition to the add-scheduler page? Like I guess it might make sense if you selected jobs from the main page and then opted to combine them into a sequence, rather than having the possible jobs populated from api/scheduler/queueable? If not, I feel like some message to the effect that selected jobs will be bundled into the sequence would be appropriate (I'm assuming that they the jobs now will be run based on the sequence's cron schedule, rather than their previous per-job schedule?)

**Comment by ismay** [comment]
Thanks for the review! So regarding your first point

> I don't see errors coming through in the UI. Do you need to subscribe to hasSubmitErrors submitErrors in the Form component in SequenceAddFormContainer? I'm a bit confused by the react-final-form [docs](https://final-form.org/docs/react-final-form/types/FormRenderProps) because they seem to suggest you need to subscribe, but the submitting prop seems to be working without subscription 🤔

Good question. I don't think that that's necessary. If I'm reading the docs correctly the default is to subscribe to all form state changes: https://final-form.org/docs/react-final-form/types/FormProps#subscription. Also it works for the other form components without an explicit subscription, so this is meant for subscribing to a smaller slice of formstate, the way I read it:

image

**Comment by ismay** [comment]
> I'm not sure if there is a plan to deal with the overall list page separately, but currently after the sequence gets added, when you click on edit, it will open the edit menu up for the first job in the sequence. Also if you click delete, it seems to just delete the first job in the sequence.

Yes the list page currently isn't really aware of queues. It works a little, but that's just a happy coincidence. I'll add the edit page next and then the list page. I figured that splitting it up that way would be easier to review, compared to one huge PR.

The route to add a queue isn't currently in the UI, so users shouldn't be adding queues anyway. So they shouldn't be encountering any weirdness with the list unless they specifically add the queue via the API. The reason for having the route be functional already even though it isn't navigable in the UI is so we can e2e test it.

**Comment by ismay** [comment]
> UI questions (so they might not be relevant if this has been decided elsewhere): I was a little confused by the fact that I had to add jobs, then I had to add a new sequence and that adding to the sequence would remove the job from the list. I don't know if this is something that would be addressed in the transition to the add-scheduler page? Like I guess it might make sense if you selected jobs from the main page and then opted to combine them into a sequence, rather than having the possible jobs populated from api/scheduler/queueable? If not, I feel like some message to the effect that selected jobs will be bundled into the sequence would be appropriate

Yeah this is what we settled on in talking it through with Jan and Joe. I'm open to making improvements though, if you feel it's weird behaviour. The queueable endpoint is meant to be an easy way to get the queueable jobs for the frontend, since you can't have one job be a part of multiple queues. Or add a queue to a queue.

Basically the idea is that single jobs have either a cron or a delay schedule. Making sure that they run in sequence is hard to do with that kind of scheduling, so for that there's now the queue. So first you configure individual jobs, and after that you can add them to a queue.

Let me know if you think the UI could be improved anywhere to make that clearer to the user.

> I'm assuming that they the jobs now will be run based on the sequence's cron schedule, rather than their previous per-job schedule?

That's right. The user can configure when the queue starts, and after that the jobs will run in order, each job starting when the preceding job has finished, until the queue is depleted.

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js:29]: Yeah this is on purpose. The idea is based on this section of the FF docs:

image
> https://final-form.org/docs/final-form/types/Config#onsubmit

So anything that is not a server-side validation error will reject the promise, as they're suggesting here. In looking at this though, I don't think that this promise rejection is dealt with anywhere. We should make sure that we display this in a noticebox. Good catch, I thought we were already handling this, but it doesn't look like it.

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js:25]: Hmm, it seems like we should be getting those. Same as we do with jobs. It's my understanding that this type of error reporting hasn't been implemented on all endpoints though, so I can imagine that this hasn't been added in the backend for this endpoint.

@jbee, do you know if the `scheduler/queues/${name}` endpoint should return `errorReports` on submission errors, similar to the `jobConfigurations` endpoint?

Comment (jbee) [reply src/hooks/job-queues/use-submit-job-queue.js:25]: There is no error report in a response. Each error has a unique error code and message for each possible error.

>   E7020( "Job queue does not exist: `{0}`" ),
>   E7021( "Job queue already exist: `{0}`" ),
>   E7022( "Job `{0}` is already used in another queue: `{1}`" ),
>   E7023( "Job `{0}` is a system job and as such cannot be added to a queue.`" ),
>   E7024( "Job queue must have at least two jobs." ),

File: src/components/FormFields/SequenceOrderField/SequenceTransferTitle.js
Comment (ismay) [reply src/components/FormFields/SequenceOrderField/SequenceTransferTitle.js]: Fixed in a3c8b45

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js:25]: I see. Is there a reason that this endpoint diverges from the way the other endpoints report back submission errors? From a frontend perspective it would be more maintainable to have a single way for the backend to report back errors. Are we moving the legacy ones to the new pattern as well? Or does it diverge for another reason?

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js:29]: Created an issue for this: https://dhis2.atlassian.net/browse/DHIS2-15556. @kabaros, what do you think? Shall we handle this across the entire app before merging?

Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js:29]: I suppose for simplicity we could just handle these errors with final-forms FORM_ERROR key. Final form does not supply a convenient mechanism for handling the errors from onSubmit promise rejections. So handling them the same way as generic form errors seems appropriate in the absence of anything else.

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js:25]: For the pattern that @tomzemp reported (trying to create an entity with a name that already exists) the backend will return a response with error reports if it's a job. So for queues we do not get errorReports for this particular pattern.

So if I understand correctly this difference in behaviour between the two endpoints is intentional?

Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js:25]: So Jan and I discussed this on slack a bit. Basically the gist is that the errorReports we're using for the jobs are just a side-effect of the specific backend operations that endpoint carries out. The new queue endpoints work differently and thus do not get the error reports.

So in short, we'll need to adapt the formatError fn to deal with the new type of error response.

File: src/hooks/job-queues/use-submit-job-queue.js
Comment (ismay) [reply src/hooks/job-queues/use-submit-job-queue.js]: I've gone ahead and went with the last approach. Should be resolved with the last 3 commits.

**Comment by ismay** [comment]
@kabaros, @tomzemp I've adjusted the error formatting to deal with all errors and display them as generic errors in the form. That should resolve both the issues that came up. Any comments on the changes? Otherwise I'll be merging this. Let me know!

**Review by tomzemp — APPROVED** [review]

**Comment by tomzemp** [comment]
> Yeah this is what we settled on in talking it through with Jan and Joe. I'm open to making improvements though, if you feel it's weird behaviour. The queueable endpoint is meant to be an easy way to get the queueable jobs for the frontend, since you can't have one job be a part of multiple queues. Or add a queue to a queue.
> 
> Basically the idea is that single jobs have either a cron or a delay schedule. Making sure that they run in sequence is hard to do with that kind of scheduling, so for that there's now the queue. So first you configure individual jobs, and after that you can add them to a queue.

I'll defer to you collectively 😄. I guess when you get around to adding the UI to add a queue from the main page, I think "add queue" option might be a bit more understandable if you had an intermediary step that asked the user to select the existing jobs they want to combine into a queue before proceeding. I just wonder if people will understand that putting a job in the queue will get rid of the existing job (with its existing cron schedule). My expectation without knowing how it works would be that if I add a job to a queue, it would continue to exist as a stand-alone job (with its previously defined cron expression) and would also exist as a step in the newly defined queue (with the queue's cron expression), but maybe that is not what others would expect.

**Comment by ismay** [comment]
> I'll defer to you collectively smile. I guess when you get around to adding the UI to add a queue from the main page, I think "add queue" option might be a bit more understandable if you had an intermediary step that asked the user to select the existing jobs they want to combine into a queue before proceeding.

Yeah definitely, that button is also part of the design, see here: https://www.figma.com/file/YVXIsd80IgZnTfMDE0qOGf/Scheduler?type=design&node-id=145-1279&mode=design

image

>  I just wonder if people will understand that putting a job in the queue will get rid of the existing job (with its existing cron schedule). My expectation without knowing how it works would be that if I add a job to a queue, it would continue to exist as a stand-alone job (with its previously defined cron expression) and would also exist as a step in the newly defined queue (with the queue's cron expression), but maybe that is not what others would expect. I just wonder if people will understand that putting a job in the queue will get rid of the existing job (with its existing cron schedule). My expectation without knowing how it works would be that if I add a job to a queue, it would continue to exist as a stand-alone job (with its previously defined cron expression) and would also exist as a step in the newly defined queue (with the queue's cron expression), but maybe that is not what others would expect.

Yeah I can imagine. Initially I also felt the process was a little unclear. I think it's probably best that we take another look at the flow when the list PR is ready. At that stage it would be good to see if the process is clear enough. I'll ping you when it's ready for review.

**Pull request merged**
